### PR TITLE
[apple-authentication] add doc comments in source

### LINKED ISSUE #12598: [AppleAuthentication][docs]  AppleAuthentication.signOutAsync() is not documented
### Summary

[AppleAuthentication.signOutAsync()](https://github.com/expo/expo/blob/3f978ec39f1cc69d9672c68dd7ae8b13083c91ff/packages/expo-apple-authentication/src/AppleAuthentication.ts#L61) is not documented in the [package's docs](https://github.com/expo/expo/blob/sdk-41/docs/pages/versions/unversioned/sdk/apple-authentication.md).

This makes it appear as though the package has no sign out function/method.

### Link to the related docs page

https://docs.expo.io/versions/latest/sdk/apple-authentication/#methods

### PR DESCRIPTION
**Pull request by Simek**
# Why

To utilize the auto generation feature for API section the package source code must contain the relevant documentation.

Refs #12239
Fixes #12598

Fixes ENG-910 ⚠️ **(check out for known issues)**

# How

This PR moves the current documentation of `AppleAuthentication` API - https://docs.expo.io/versions/latest/sdk/apple-authentication/ - (with minor tweaks and addition)  to the package source code, with backport to v41 SDK docs.

There are some parts with missing docs, partially missing docs and unverified (not published before) comments, but overall the content looks better/more complete than the current docs. If anyone want to contribute those, before the merge please leave the change suggestion.

✨ This PR also includes the following changes:
* [tools] better support for multiple entries per module
* [docs] use more generic `Link` component for MD rendering
* [docs] fix missing fragment key warning

# Test Plan

The changes and generated content have been tests running docs website on `localhost`.

# Preview

Screenshot 2021-04-21 at 18 56 50

# Checklist

- [X] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [X] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [X] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

CC: @bbarthec @brentvatne

### CODE REVIEW

**Review by esamelson — APPROVED** [review]
docs migration looks good to me, just one small comment/question!

File: docs/pages/versions/v41.0.0/sdk/apple-authentication.md [hunk 1]
```diff
@@ -77,80 +78,6 @@ Apple's response includes a signed JWT with information about the user. To ensur
...
-
-- **requestedScopes (_[AppleAuthenticationScope](#appleauthenticationappleauthenticationscope)[]_)** (optional) - Array of user information scopes to which your app is requesting access. Note that the user can choose to deny your app access to any scope at the time of logging in. You will still need to handle `null` values for any scopes you request. Additionally, note that the requested scopes will only be provided to you **the first time each user signs into your app**; in subsequent requests they will be `null`.
-- **state (_string_)** (optional) - An arbitrary string that is returned unmodified in the corresponding credential after a successful authentication. This can be used to verify that the response was from the request you made and avoid replay attacks. More information on this property is available in the OAuth 2.0 protocol [RFC6749](https://tools.ietf.org/html/rfc6749#section-10.12).
-- **nonce (_string_)** (optional) - An arbitrary string that is used to prevent replay attacks. See more information on this in the [OpenID Connect specification](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowSteps).
```
[the remark this answers is not in the document]
Comment (Simek) [reply docs/pages/versions/v41.0.0/sdk/apple-authentication.md:134]: That's a good catch, thank you! Part of comments were already in the source, I forgot to update the descriptions of `AppleAuthenticationSignInOptions` fields, to match the descriptions in docs. Will update the PR soon. 🙂

[the remark this answers is not in the document]
Comment (Simek) [reply docs/pages/versions/v41.0.0/sdk/apple-authentication.md:134]: Screenshot 2021-04-21 at 21 59 47

File: packages/expo-apple-authentication/src/AppleAuthentication.ts [hunk 2]
```diff
@@ -1,23 +1,47 @@
...
+ *
+ * Additionally, you will only receive Apple Authentication Credentials the first time users sign
+ * into your app, so you must store it for later use. It's best to store this information either
+ * server-side, or using [SecureStore](./securestore), so that the data persists across app installs.
+ * You can use [`AppleAuthenticationCredential.user`](#appleauthenticationcredential) to identify
+ * the user, since this remains the same for apps released by the same developer.
+ *
+ * @param options An optional [`AppleAuthenticationSignInOptions`](#appleauthenticationsigninoptions) object
+ * @return A promise that fulfills with an [`AppleAuthenticationCredential`](#appleauthenticationcredential)
+ * object after a successful authentication, and rejects with ERR_CANCELED if the user cancels the
```
Comment (brentvatne) [inline packages/expo-apple-authentication/src/AppleAuthentication.ts:42]:
```suggestion
 * object after a successful authentication, and rejects with `ERR_CANCELED` if the user cancels the
```

File: packages/expo-apple-authentication/src/AppleAuthentication.types.ts [hunk 3]
```diff
@@ -192,38 +206,63 @@ export enum AppleAuthenticationCredentialState {
...
 }
 
+// @needsAudit
 /**
- * Controls the predefined style of the authenticating button.
+ * An enum whose values control which pre-defined color scheme to use when rendering an [`AppleAuthenticationButton`](#appleauthenticationappleauthenticationbutton).
  */
 export enum AppleAuthenticationButtonStyle {
+  /**
+   * White button with black text
```
Comment (brentvatne) [inline packages/expo-apple-authentication/src/AppleAuthentication.types.ts:257]: in `AppleAuthenticationUserDetectionStatus` the comments all ended with a `.` - should we do that here too?

**Review by brentvatne — APPROVED** [review]

File: tools/src/commands/GenerateDocsAPIData.ts [hunk 4]
```diff
@@ -50,12 +54,25 @@ const executeCommand = async (
...
     await app.generateJson(project, jsonOutputPath);
+    const output = await fs.readJson(jsonOutputPath);
+
+    if (Array.isArray(entryPoint)) {
+      const filterEntries = entryPoint.map(entry => entry.substring(0, entry.lastIndexOf('.')));
+      output.children = output.children
+        .filter(entry => filterEntries.includes(entry.name))
+        .map(entry => entry.children)
+        .flat()
+        .sort((a, b) => (a.name > b.name) ? 1 : ((b.name > a.name) ? -1 : 0));
```
Comment (bbarthec) [inline tools/src/commands/GenerateDocsAPIData.ts:65]:
```suggestion
        .sort((a, b) => a.localeCompare(b));
```

Comment (tsapeta) [reply tools/src/commands/GenerateDocsAPIData.ts:65]: rather this way 👇 
```suggestion
        .sort((a, b) => a.name.localeCompare(b.name));
```

**Review by tsapeta — APPROVED** [review]
Looks good to me, except the comment I put under Bartek's review 😉

**Comment by Simek** [comment]
Sort update pushed, thank you! I see no one is keen to fill out the missing parts, hope it's not a blocker. Also still got no response on the link issue, hope it's fine and won't cause problem on deployed version.

**Review by bbarthec — APPROVED** [review]

**Pull request merged**